binary_search_tree/tree.py

```diff
@@ -7,52 +7,174 @@ def __init__(self, key, val = None):
+        if curr_node == None:
+            return None
+        
+        if key < curr_node.key:
+            curr_node.left = self.find_helper(curr_node.left, key, value)
+        else:
+            curr_node.right = self.find_helper(curr_node, key, value)
+        return curr_node
+    
+    # Time Complexity: O(log n) *if balanced
+    # Space Complexity: O(log n) *if balanced
+    def find_recursive(self, key, value = None):
+        if self.root == None:
+            return None
+        else:
+            self.find_helper(self.root, key, value)
+        
+    def preorder_helper(self, curr_node, trav_list):
+        if curr_node == None:
+            return 
+        else:
+            trav_list.append({"key": curr_node.key, "value": curr_node.value})
+            self.preorder_helper(curr_node.left, trav_list)
+            self.preorder_helper(curr_node.right, trav_list)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)    
+    def preorder(self):
+        if self.root == None:
+            return []
+        traversal_list = []
+        self.preorder_helper(self.root, traversal_list)
+        return traversal_list
+
+    def inorder_helper(self, curr_node, trav_list):
+        if curr_node == None:
+            return
+        else:
+            self.inorder_helper(curr_node.left, trav_list)
+            trav_list.append({"key": curr_node.key, "value": curr_node.value})
+            self.inorder_helper(curr_node.right, trav_list)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        if self.root == None:
+            return []
+        traversal_list = []
+        self.inorder_helper(self.root, traversal_list)
+        return traversal_list
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def preorder(self):
-        pass
+    def postorder_helper(self, curr_node, trav_list):
+        if curr_node == None:
+            return
+        else:
+            self.postorder_helper(curr_node.left, trav_list)
+            self.postorder_helper(curr_node.right, trav_list)
+            trav_list.append({"key": curr_node.key, "value": curr_node.value})
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) 
     def postorder(self):
-        pass
+        if self.root == None:
+            return []
+        traversal_list = []
+        self.postorder_helper(self.root, traversal_list)
+        return traversal_list
+
+    def height_helper(self, curr_node):
+        if curr_node == None:
+            return 0
+        left_height = self.height_helper(curr_node.left)
+        right_height = self.height_helper(curr_node.right)
+        return max(left_height, right_height) + 1
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def height(self):
```

Comment:
👍 The space complexity is O(h) where h is the height of the tree, but you're right if it's unbalanced h could be the number of nodes.